### PR TITLE
Fix grpc trace bin headers to use base64 encode/decode.

### PR DESCRIPTION
**Pull request by bogdandrutu**

### CODE REVIEW

**Comment by bogdandrutu** [comment]
endpoints

File: src/api_manager/context/request_context.cc [hunk 1]
```diff
@@ -146,10 +147,17 @@ RequestContext::RequestContext(std::shared_ptr<ServiceContext> service_context,
...
-      request_->FindHeader(kCloudTraceContextHeader, &trace_context_header);
+    if (!request_->FindHeader(kCloudTraceContextHeader,
+                              &trace_context_header)) {
+      // Grpc binary headers are base64 encoded, decode the header before
+      // parsing it.
+      std::string base64_trace_context_header;
+      if (request_->FindHeader(kGRpcTraceContextHeader,
+                               &base64_trace_context_header)) {
+        header_type = HeaderType::GRPC_TRACE_CONTEXT;
+        absl::Base64Unescape(base64_trace_context_header,
```
Comment (qiwzhang) [inline src/api_manager/context/request_context.cc]: Need to handle decode error.

File: src/api_manager/context/request_context.cc [hunk 2]
```diff
@@ -376,15 +384,30 @@ const std::string RequestContext::FindClientIPAddress() {
...
-  // The span id in the header will be the backend span's id.
-  std::string trace_context = cloud_trace()->ToTraceContextHeader(
-      backend_span_->trace_span()->span_id());
+  // TODO: A better logic would be to send for GRPC backends the grpc-trace-bin
+  // header, and for http/https backends the X-Cloud-Trace-Context header.
+
+  std::string trace_context_header;
+  if (cloud_trace()->header_type() == HeaderType::GRPC_TRACE_CONTEXT) {
+    // The span id in the header will be the backend span's id.
+    std::string base64_trace_context_header =
```
Comment (qiwzhang) [inline src/api_manager/context/request_context.cc]: this name is confusing.  It is raw_,  not base64 encoded.

File: src/api_manager/context/request_context.cc [hunk 3]
```diff
@@ -376,15 +384,30 @@ const std::string RequestContext::FindClientIPAddress() {
...
+        cloud_trace()->ToTraceContextHeader(
+            backend_span_->trace_span()->span_id());
+    // For grpc the header must be base64 encoded because this is a binary
+    // header.
+    absl::Base64Escape(base64_trace_context_header, &trace_context_header);
```
Comment (qiwzhang) [inline src/api_manager/context/request_context.cc]: I prefer to have base64 encode decoding inside cloud_trace class, based on the type.

File: src/api_manager/context/request_context.cc (hunk 2 above)
Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Actually is not raw, the base64 comes from grpc/esp integration. If I use directly the grpc::ClientContext I will but the non-base64 (raw) value, but the gRPC protocol does base64 encoding and you keep that inside esp (in the proxy code for example you explicitly base64 encode all the binary headers).

**Comment by qiwzhang** [comment]
@bogdandrutu I am going to release a new ESP this week.   Please hurry up this PR if you want the new ESP to have this change.

File: src/api_manager/context/request_context.cc (hunk 3 above)
Comment (qiwzhang) [inline src/api_manager/context/request_context.cc]: Can we put calling Base64Escape inside ToTraceContextHeader if it is grpc_trace type

Do the same for CreateCloudTrace.  if grpc_trace type,  calling Base64Unscape.

Comment (qiwzhang) [reply src/api_manager/context/request_context.cc]: In this way, the code is cleaner in request_context.cc

File: src/api_manager/context/request_context.cc (hunk 1 above)
Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Done.

File: src/api_manager/context/request_context.cc (hunk 2 above)
Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Done, renamed and moved.

File: src/api_manager/context/request_context.cc (hunk 3 above)
Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Done.

Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: Hopefully with the move everything is fine.

**Comment by qiwzhang** [comment]
could you also rebase your changes

File: src/api_manager/context/request_context.cc [hunk 4]
```diff
@@ -376,15 +378,18 @@ const std::string RequestContext::FindClientIPAddress() {
 void RequestContext::StartBackendSpanAndSetTraceContext() {
   backend_span_.reset(CreateSpan(cloud_trace_.get(), "Backend"));
 
-  // The span id in the header will be the backend span's id.
-  std::string trace_context = cloud_trace()->ToTraceContextHeader(
+  // TODO: A better logic would be to send for GRPC backends the grpc-trace-bin
+  // header, and for http/https backends the X-Cloud-Trace-Context header.
```
Comment (qiwzhang) [inline src/api_manager/context/request_context.cc]: This is a good idea.   Actually we have the data

It is here:  https://github.com/cloudendpoints/esp/blob/master/src/api_manager/context/request_context.cc#L296

Comment (qiwzhang) [reply src/api_manager/context/request_context.cc]: It should be an easy change.  Could you make this change in this PR

Comment (bogdandrutu) [reply src/api_manager/context/request_context.cc]: I prefer to not do it in a rush right now, I would send you a different PR.

**Comment by bogdandrutu** [comment]
Done with the rebase.

**Pull request merged**